### ISSUE TITLE
rms_normalization and layer_normalization truncate integer and bool inputs to the input dtype

### ISSUE THREAD

**Issue by JyotinderSingh** [keras-team/keras#23708]
### Description

`keras.ops.rms_normalization` and `keras.ops.layer_normalization` compute in float32 and then cast the result back to the dtype of the input. For an integer or bool input, this cast truncates the normalized values. The functions return wrong values with no error on the torch, jax, tensorflow and numpy backends. `RMSNorm.compute_output_spec` and `LayerNorm.compute_output_spec` report the input dtype, so the graph shows the same wrong dtype.

`keras.ops.normalize` had the same problem. #23565 corrects it there: a float input keeps its dtype, and an integer or bool input gets a float output dtype.

### Reproduction

```python
import numpy as np

import keras

x = np.array([[1, 2, 3], [4, 5, 6]], dtype="int32")
print(keras.ops.rms_normalization(x))
print(keras.ops.layer_normalization(x))
print(keras.ops.rms_normalization(np.array([[True, False, True]])))
```

Output on the torch, jax, tensorflow and numpy backends:

```
int32 [[0, 0, 1], [0, 0, 1]]
int32 [[-1, 0, 1], [-1, 0, 1]]
bool  [[True, False, True]]
```

The same calls with a float32 input give the correct values:

```
[[0.463, 0.926, 1.389], [0.790, 0.987, 1.184]]
[[-1.225, 0.0, 1.225], [-1.225, 0.0, 1.225]]
[[1.225, 0.0, 1.225]]
```

The layers show the same effect. `keras.layers.RMSNormalization()(np.arange(6).reshape(2, 3))` returns int32 `[[0, 0, 1], [0, 0, 1]]`. `keras.layers.LayerNormalization()` on the same input returns float32 `[[-1, 0, 1], [-1, 0, 1]]`. The multiplication with `gamma` changes the dtype to float32, but the values stay truncated.

### Cause

In `keras/src/ops/nn.py` at 4749f97e8:

- `_rms_normalization` (lines 3044 to 3073) sets `original_dtype = backend.standardize_dtype(x.dtype)` and returns `backend.ops.cast(outputs, original_dtype)` on the fused path and on the fallback path.
- `_layer_normalization` (lines 3160 to 3217) does the same on its three return paths.
- `RMSNorm.compute_output_spec` (line 2993) and `LayerNorm.compute_output_spec` (line 3084) return `KerasTensor(shape=x.shape, dtype=x.dtype)`.

`backend.result_type(x.dtype, "float32")` gives float32 for an integer input, so the calculation itself is correct. The cast back to `original_dtype` then truncates the result.

### Possible fix

Apply the rule from #23702 and from the `_promote_to_float` helper in #23545. A float input keeps its dtype. An input that is not a float gets `backend.floatx()`:

```python
    original_dtype = backend.standardize_dtype(x.dtype)
    if backend.is_float_dtype(original_dtype):
        output_dtype = original_dtype
    else:
        output_dtype = backend.floatx()
```

Use `output_dtype` for each final cast and in the two `compute_output_spec` methods. Add `int32` and `bool` cases to `test_rms_normalization` and `test_layer_normalization` in `NNOpsDtypeTest`.

Do not use `backend.result_type(x.dtype, float)` for this. It gives float16 under `floatx="bfloat16"` (see #23712), and it gives float32 under `floatx="float64"` on the torch and numpy backends, where `floatx()` gives float64. #23565 uses `backend.result_type(x.dtype, backend.floatx())` for `normalize`. That call also gives float32 under `floatx="float64"` on those backends, so `normalize` needs the same change as this issue.

### Environment

- Keras version: master at 4749f97e8
- Backends: torch 2.10.0, jax 0.9.1, tensorflow 2.20.0, numpy
- Python version: 3.12
- OS: macOS (Darwin 25.6.0)

**Comment by JyotinderSingh** [comment]
The possible fix in the issue text now keeps complex dtypes, casts the `RMSNormalization` output to `compute_dtype`, and tests `floatx` values other than float32. Without the layer cast, `RMSNormalization(dtype="mixed_float16")` on an `int32` input gives a float32 eager output and a float16 symbolic output. Under `floatx="float64"`, the `result_type(x.dtype, floatx())` rule gives float32 on the numpy, torch and openvino backends.

What each open PR must change:

- #23711 (`rms_normalization`, `layer_normalization`):
  - In `_normalization_output_dtype`, return `backend.floatx()` for integer and bool input, not `backend.result_type(dtype, backend.floatx())`.
  - Keep complex dtypes: `backend.is_float_dtype(dtype) or "complex" in dtype`. Without this check, a `complex64` input loses its imaginary part.
  - Cast the result of `RMSNormalization.call` to `self.compute_dtype`, as `LayerNormalization.call` does.
  - In the new tests, use `backend.floatx()` as the expected dtype.
  - Add `test_normalization_non_float_input_floatx` (`bfloat16`, `float16`, `float64`) and the two layer tests `test_non_float_input` from the issue text.
- #23565 (`normalize`):
  - Apply the `backend.floatx()` rule from the last review in `_normalize`, `Normalize.compute_output_spec` and the test.
  - Keep complex dtypes with the same check. On `master`, `order=1` and `order=3` give the correct complex64 result on the torch, jax and numpy backends.
  - Cast the result of `UnitNormalization.call` to `self.compute_dtype`.
  - Rebase onto `master` to remove the conflict in `_normalize`.

The compute dtype `result_type(x.dtype, "float32")` raises for float8 and computes in float32 for integer input under `floatx="float64"`. This scope moved to #23751 (item 8). #23754 adds the cast of integer and bool inputs at the start of `call` in the normalization layers.

**Remaining work under this issue**

- [ ] #23711: return `backend.floatx()` for integer and bool input.
- [ ] #23711: keep complex dtypes in `_normalization_output_dtype`.
- [ ] #23711: cast the `RMSNormalization.call` result to `self.compute_dtype`.
- [ ] #23711: add the `floatx` test and the two layer tests.
- [ ] #23565: apply the same rule and complex check in `_normalize` and its spec.
- [ ] #23565: cast the `UnitNormalization.call` result to `self.compute_dtype`.
- [ ] #23565: rebase onto `master`.
- [ ] Put item 8 of #23751 into #23711, or leave it for PR 2 of that issue.
